### PR TITLE
Stake contract

### LINKED ISSUE #148: Implement Stake Contract

### PR DESCRIPTION
**Pull request by vlopes11**
Resolves #148

### CODE REVIEW

File: contracts/stake/src/wasm.rs [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+                PublicKey,
+                Signature,
+                Note,
+                Vec<u8>,
+            ) = Canon::decode(&mut source).expect("Failed to parse arguments");
+
+            contract.withdraw(pk, signature, note, withdraw_proof);
+        }
+
+        TX_SLASH => panic!("The arbitrationContract is yet to be defined."),
```
Comment (ZER0) [inline contracts/stake/src/wasm.rs:79]: The slash should be still implemented at least for the branch where the `arbitrationContract` is not involved (see: https://app.gitbook.com/@dusk-network/s/specs/specifications/smart-contracts/genenesis-contracts/stake-contract-1/).

Edit: I spoke with @toghrulmaharramov and the idea would be ignore the `if` at all currently and burn always the token (the `else` in the documentation).

**Review by ZER0 — CHANGES_REQUESTED** [review]
Overall looks good, but there are few changes to address. Most notably, the `slash` method needs to be implemented (without `arbitrationContract`), and the `getStake` should be implemented as well.

Check with @toghrulmaharramov if "spurious" in this case means native (as `findNotes`) or not (since we discovered "spurious" had a different meaning, and we should update the docs too)

File: contracts/stake/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+use dusk_hamt::Map;
+use dusk_pki::PublicKey;
+use rusk_abi::PaymentInfo;
+
+#[cfg(target_arch = "wasm32")]
+mod wasm;
+
+pub const EPOCH: u32 = 2160;
+pub const MATURITY: u32 = 2 * EPOCH;
+pub const VALIDITY: u32 = 56 * EPOCH;
```
[the remark this answers is not in the document]
Comment (vlopes11) [reply contracts/stake/src/lib.rs]: Done

File: contracts/stake/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+#![no_std]
+
+extern crate alloc;
+
+use canonical_derive::Canon;
+use dusk_abi::ContractId;
+use dusk_bytes::Serializable;
+use dusk_hamt::Map;
+use dusk_pki::PublicKey;
+use rusk_abi::PaymentInfo;
```
[the remark this answers is not in the document]
Comment (vlopes11) [reply contracts/stake/src/lib.rs]: Done

File: contracts/stake/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+#[cfg(target_arch = "wasm32")]
+mod wasm;
+
+pub const EPOCH: u32 = 2160;
+pub const MATURITY: u32 = 2 * EPOCH;
+pub const VALIDITY: u32 = 56 * EPOCH;
+pub const MINIMUM_STAKE: u64 = 10_000;
+pub const MAXIMUM_STAKE: u64 = 1_000_000;
+pub const PAYMENT_INFO: PaymentInfo = PaymentInfo::Transparent(None);
```
[the remark this answers is not in the document]
Comment (vlopes11) [reply contracts/stake/src/lib.rs]: Done

File: contracts/stake/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+pub type PublicKeyBytes = [u8; PublicKey::SIZE];
```
[the remark this answers is not in the document]
Comment (vlopes11) [reply contracts/stake/src/lib.rs]: As discussed, the PublicKey does not implement the requirements of dusk-hamt (eg `Eq`) so its impossible to use the native type.

However, the next version of dusk-pki should contain this change

File: contracts/stake/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    }
+
+    pub fn extend(&mut self) {
+        self.expiration += VALIDITY;
+    }
+}
+
+#[derive(Debug, Default, Clone, Canon)]
+pub struct StakeContract {
+    staked: Map<PublicKeyBytes, Stake>,
```
[the remark this answers is not in the document]
Comment (vlopes11) [reply contracts/stake/src/lib.rs]: Ditto

File: contracts/stake/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+impl StakeContract {
+    pub fn transfer_contract() -> ContractId {
+        ContractId::from([
+            0xd3, 0xf8, 0x7f, 0xfc, 0x1b, 0xc7, 0x43, 0x1d, 0xde, 0x81, 0x5f,
+            0xb1, 0xe1, 0x1b, 0xd0, 0xfe, 0x88, 0x37, 0x1a, 0x15, 0x4a, 0xec,
+            0x27, 0x5d, 0xed, 0x2, 0x4d, 0x8c, 0xc0, 0xf7, 0x99, 0x5f,
+        ])
+    }
```
[the remark this answers is not in the document]
Comment (vlopes11) [reply contracts/stake/src/lib.rs]: Done

File: contracts/stake/src/wasm.rs (hunk 1 above)
Comment (vlopes11) [reply contracts/stake/src/wasm.rs]: Done

File: contracts/stake/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+use canonical_derive::Canon;
+use dusk_bytes::Serializable;
+use dusk_hamt::Map;
+use dusk_pki::PublicKey;
+
+#[cfg(target_arch = "wasm32")]
+mod wasm;
+
+pub const MATURITY: u32 = 2 * rusk_abi::EPOCH;
+pub const VALIDITY: u32 = 56 * rusk_abi::EPOCH;
```
Comment (CPerezz) [inline contracts/stake/src/lib.rs:20]: This should be moved to rusk-abi since it's also part of `bid contract`

**Review by CPerezz — APPROVED** [review]
LGTM once the comment is addressed.

We might want to add a CHANGELOG so that we can keep track of the history of this member easily.

**Review by ZER0 — APPROVED** [review]

**Pull request merged**